cid-decode/src/main.rs

```diff
@@ -53,51 +62,136 @@ fn main() -> Result<()> {
         let entry = results.entry(res.clone()).or_default();
         *entry += 1;
 
-        s.clear();
+        buffer.clear();
     }
     // Restore panic hook.
-    panic::set_hook(h);
+    panic::set_hook(default_panic_hook);
 
     // We only get here if reading from Stdin fails...
-    rdr.read_line(&mut s)?;
+    rdr.read_line(&mut buffer)?;
 
     Ok(())
 }
 
 #[derive(Debug, Clone)]
 struct Metadata {
-    base: multibase::Base,
+    base: cid::multibase::Base,
     version: cid::Version,
-    codec: cid::Codec,
-    hash: Option<multihash::Code>,
+    codec: self::codec::Codec,
+    hash: Option<cid::multihash::Code>,
     hash_len: usize,
 }
 
-fn do_single(s: &str) -> Result<Metadata> {
-    let c = cid::Cid::try_from(s)?;
-    if c.version() == cid::Version::V0 {
-        return Ok(Metadata {
-            base: multibase::Base::Base58Btc,
-            version: c.version(),
-            codec: c.codec(),
-            hash: match std::panic::catch_unwind(|| c.hash().algorithm()) {
-                Ok(h) => Some(h),
-                Err(_) => None,
-            },
-            hash_len: c.hash().digest().len(),
-        });
-    }
-
-    let (b, _) = multibase::decode(s)?;
-
-    Ok(Metadata {
-        base: b,
+fn do_single(line: &str) -> Result<Metadata> {
+    let c = cid::Cid::try_from(line)?;
+    return Ok(Metadata {
+        base: if c.version() == cid::Version::V0 {
+            cid::multibase::Base::Base58Btc
+        } else {
+            cid::multibase::decode(line)?.0
+        },
         version: c.version(),
-        codec: c.codec(),
-        hash: match std::panic::catch_unwind(|| c.hash().algorithm()) {
-            Ok(h) => Some(h),
+        codec: codec::Codec::try_from(c.codec())?,
+        hash: match std::panic::catch_unwind(|| c.hash().code()) {
+            Ok(h) => Some(cid::multihash::Code::try_from(h)?),
             Err(_) => None,
         },
         hash_len: c.hash().digest().len(),
-    })
+    });
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Very nice!
We should make this a table-driven test, i.e.: 
1. Pick a few notable CIDs with various versions, codecs, etc.,
2. Put them in a table, something like (input CID, expected result)
3. Throw them all through the tests

As an example, take something like this (which is Go, but illustrates the idea): https://github.com/chihaya/chihaya/blob/main/frontend/http/bencode/encoder_test.go#L12-L46 and turn it into Rust, I guess :)
There's probably also examples of doing this in Rust. I found this, which uses a macro, but maybe there's others as well: https://github.com/BurntSushi/fst/blob/master/src/raw/tests.rs#L62-L100
We could, e.g., use `lazy_static` to construct a hashmap of test cases, which I think is pretty clean. Probably need to try some things out and see what sticks...

Reply:
We can do that but I whould rather spend my time on the `bitswap-monitor-client` :).
I think more extensive tests are more of a nice to have and should not block this PR. 

Also the cid-decode code is rather small. Maybe our tests now are sufficient.